embedded-graphics/src/primitives/triangle/fill_scanline_iterator.rs

```diff
@@ -0,0 +1,203 @@
+//! A scanline iterator that returns every point in a triangle once.
+use crate::{
+    geometry::Point,
+    primitives::{
+        line::{self, Line},
+        triangle::{sort_two_x, sort_yx, Triangle},
+        Primitive,
+    },
+};
+
+/// Iterator over all points inside the triangle.
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
+pub struct FillScanlineIterator {
+    /// Left-most edge of the triangle
+    line_a: line::Points,
+
+    /// Right-most edge of the triangle
+    line_b: line::Points,
+
+    /// Bottom edge of the triangle
+    line_c: line::Points,
+
+    /// Horizontal line
+    scan_points: line::Points,
+
+    /// The first point of the ab edge in the next line
+    next_a: Option<Point>,
+
+    /// The first point of the c edge in the next line
+    next_c: Option<Point>,
+}
+
+impl FillScanlineIterator {
+    pub(in crate::primitives) fn new(triangle: &Triangle) -> Self {
+        let (v1, v2, v3) = sort_yx(triangle.p1, triangle.p2, triangle.p3);
+
+        let line_a = if v1.y == v2.y {
+            line::Points::empty()
+        } else {
+            Line::new(v1, v2).points()
+        };
+        let line_b = if v2.y == v3.y {
+            line::Points::empty()
+        } else {
+            Line::new(v2, v3).points()
+        };
+        let line_c = Line::new(v1, v3).points();
+
+        Self {
+            line_a,
+            line_b,
+            line_c,
+            next_a: None,
+            next_c: None,
+            scan_points: line::Points::empty(),
+        }
+    }
+
+    fn update_ab(&mut self) -> Option<(Point, Point)> {
+        let mut next_a = self
+            .next_a
+            .take()
+            .or_else(|| self.line_a.next().or_else(|| self.line_b.next()))?;
+        let first = next_a;
+        while let Some(a) = self.line_a.next().or_else(|| self.line_b.next()) {
+            if a.y == next_a.y {
+                next_a = a;
+            } else {
+                self.next_a = Some(a);
+                return Some((first, next_a));
+            }
+        }
+        Some((first, next_a))
+    }
+
+    fn update_c(&mut self) -> Option<(Point, Point)> {
+        let mut next_c = self.next_c.take().or_else(|| self.line_c.next())?;
+        let first = next_c;
+        while let Some(c) = self.line_c.next() {
+            if c.y == next_c.y {
+                next_c = c;
+            } else {
+                self.next_c = Some(c);
+                return Some((first, next_c));
+            }
+        }
+        Some((first, next_c))
+    }
+
+    fn next_scanline(&mut self) -> bool {
+        let a = self.update_ab();
+        let c = self.update_c();
+
+        match (a, c) {
+            (Some((fa, la)), Some((fc, lc))) => {
```

Comment:
Does `f` stand for `first` and `l` for `last` her? It would be better to use longer, more descriptive variable names IMO.

Can you also add a short doc comment description to this method explaining what it does? I think I understand, but a few words would help me and others.

Reply:
Yeah this is not nice :)